Send people doing *library* stabilizations over to the std-dev-guide (#1317)
rust-lang/rustc-dev-guide: src/stabilization_guide.md

Note: Before we stabilize any feature, it's the rule that it
should appear in the documentation.

### Determining the stabilization version

The version in which the feature will be stabilized *must* match
the value of [the `src/version` file in `master`] when the PR is merged.

It's worth checking [the version schedule on the Forge] to see whether
changes are coming soon.  You'll usually use the version labelled "Nightly".

> That's two versions higher than the current stable release,
> as what's currently in beta will be the next stable release,

Reviewer: ```suggestion
"Nightly" is two versions higher than the current stable release,
since what's currently in beta will be the next stable release,
```

> and any change you're making now will be in the one after that.

Reviewer: ```suggestion
and any changes you're making now will be in the one after that.
```

if something { /* XXX */ }
```

## Stabilization PR for Library Features

Once we have decided to stabilize a feature, we need to have
a PR that actually makes that stabilization happen. These kinds
of PRs are a great way to get involved in Rust, as they're typically
small -- just updating attributes.

Here is a general guide to how to stabilize a feature --
every feature is different, of course, so some features may
require steps beyond what this guide talks about.

### Update the stability attributes on the items

Library items are marked unstable via the `#[unstable]` attribute, like this:

```rust,ignore
#[unstable(feature = "total_cmp", issue = "72599")]
pub fn total_cmp(&self, other: &Self) -> crate::cmp::Ordering { ... }
```

You'll need to change that to a `#[stable]` attribute with a version:

```rust,ignore
#[stable(feature = "total_cmp", since = "1.61.0")]
```

Note that, the version number is updated to be the version number
of the stable release where this feature will appear. This can be
found by consulting [the forge](https://forge.rust-lang.org/#current-release-versions).

> Specifically, you'll want to use the version labelled "Nightly".

Reviewer 2: Although this is a perfectly fine way to check the version, it might be more reliable to consult [`src/version`](https://github.com/rust-lang/rust/blob/master/src/version) on master.  That can help avoid confusion with the version bump that happens on the Friday before release.

I might also suggest *predicting* the version that it will land in.  If it looks like it might land around the Friday before release, then there is risk that it will miss the release window.

Finally, I might suggest putting this discussion of determining the version into a dedicated section and removing the duplicate discussion above with lang stabilizations. For the most part, the process is the same between them. (And the lang one is not written as clearly.)

It is not uncommon that PRs are approved with the wrong version in them, and then there is confusion if it isn't caught, or triggers busywork for backports and such.  It would be nice to be a little more careful about that.

```

[rust-lang/rust#32409]: https://github.com/rust-lang/rust/issues/32409
[the stabilization chapter of the std dev guide]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html
[the `src/version` file in `master`]: https://github.com/rust-lang/rust/blob/master/src/version
[the version schedule on the Forge]: https://forge.rust-lang.org/#current-release-versions

> [the Friday before]: https://forge.rust-lang.org/release/process.html

Reviewer 3: These link reference names (e.g., `the stabilization chapter of the std dev guide`) are long and prone to breaking if wording is changed. Could you change them to be pithier?

E.g.: `std-guide-stabilization`, `src-version`, `release-process`.

(You'll have to add manual `[...]` link references to the text too.)

Author: Good idea.  I've added pithy names to these, and added the `(...)` link references.

(The `[...]` was a typo, right?  My syntax highlighter thinks this is right.)

Reviewer 3: No, it should be `[...]`. For example, `[look at this](src-version)` will link to `src-version` as if it is a URL. On the other hand, `[look at this][src-version]` will use the URL value of the definition `[src-version]: https://example.com` as the link.

Author: Oh!  I see; thank you.  Fixed.

Author: And I checked the rendered <https://github.com/rust-lang/rustc-dev-guide/blob/library-stabilization/src/stabilization_guide.md> this time to confirm, which I should have done last time but didn't think of.
